fix: npm-version does not git-commit nor git-tag when package in subdirectory

fernandopasik opened the pull request:
<!-- What / Why -->
<!-- Describe the request in detail. What it does and why it's being changed. -->

libnpmversion only checks one same dir if is git repo, instead use `git.find` to check parents as well

## References
fixes #2010

fernandopasik committed:
fix: look parent dirs to detect git repo on version

fernandopasik commented:
Have a problem when runing the tests, as the temp files are within this repo and that errors when using `git.find`
https://github.com/npm/cli/blob/latest/workspaces/libnpmversion/test/version.js#L339-L355

fernandopasik commented:
Tap creates folders within the git repo and that makes the test fail, I tried ignoring the folder created but `git.find` won't stop if you are within an ignored folder, I wonder if that would be required anyway for that tool

fritzy reviewed (commented):
A test to verify the new behavior may be prudent.

lukekarrys reviewed (changes requested):
@fernandopasik Can you add a test case for this behavior?

fernandopasik commented:
> @fernandopasik Can you add a test case for this behavior?

yes, I'll add the test case. I'm trying to understand why other tests have failed.

timothybonci commented:
> 

git is mocked in the tests - we don't use the package there. 
`'@npmcli/git': gitMock`

timothybonci committed:
Mock find and test with subdirectories

timothybonci commented:
@fernandopasik I opened a PR into yours with what I hope are the right changes for the tests.  I'm a GitHub noob, so this may have been a dumb way of doing it, but you should be able to see the ideas.  If this stays idle for a few days, I can open up my own MR into this repo with both our changes.

fernandopasik commented:
Sure @timothybonci can you link here your PR so we can merge?

fernandopasik committed:
Merge pull request #1 from timothybonci/patch-1

Mock find and test with subdirectories

fernandopasik committed:
refactor: use let instead of var

fernandopasik commented:
@timothybonci I've merged your changes, thanks!

fernandopasik committed:
style: use single quotes

wraithgar commented on workspaces/libnpmversion/test/version.js:
```diff
@@ -50,13 +54,21 @@ t.test('test out bumping the version in all the ways', async t => {
       actionLog.length = 0
     })
-    const path = `${dir}/git`
+    var path = `${dir}/git`
```
I don't think this is correct.

wraithgar replied:
```suggestion
    let path = `${dir}/git`
```

timothybonci replied:
Fernando made that change already, but I'm curious, since I don't do JS often, so I'm not sure on the styles.  Because it's redeclared later, should we still use let because we're not using strict?  I assumed that coding for strict was preferable.

fernandopasik replied:
reverted my commit

wraithgar commented on workspaces/libnpmversion/test/version.js:
```diff
@@ -238,7 +252,7 @@ t.test('test out bumping the version in all the ways', async t => {
       actionLog.length = 0
     })
-    const path = `${dir}/not-git`
+    var path = `${dir}/not-git`
```
```suggestion
    let path = `${dir}/not-git`
```

fernandopasik replied:
reverted the commit

timothybonci commented:
How do we get the tests re-run?  It looks like they only run once on PR open, but not on subsequent pushes, at least automatically.  Are we waiting on a manual action by a maintainer?

fernandopasik committed:
Revert "refactor: use let instead of var"

This reverts commit 2e40ec9e52f02caca5d4c15e17a8d99fe7164b5c.

timothybonci commented:
I see failed tests from yesterday - it's not obvious to me why they're failing.  I cloned the repo and can try and debug them, but I may not get to it until next week.

timothybonci committed:
use same directory in the setup and the tests

Fixing my silly mistake.  This fixes all but the last test.

timothybonci committed:
Fix final test

Last test needs subdirectory evaluation

timothybonci commented:
Just needed a good night's sleep.  https://github.com/fernandopasik/cli/pull/2 fixes the tests.

fernandopasik committed:
Merge pull request #2 from timothybonci/patch-2

use same directory in the setup and the tests

timothybonci commented:
@lukekarrys Is this waiting on you now?

timothybonci commented:
I see the new failures, thanks.

ruyadorno commented:
I'm curious on how this will interact with **workspaces**, I believe it's important to make sure this change is not going to enable every `npm version -w <ws-name>` to commit/tag, at least not until https://github.com/npm/rfcs/issues/570 is ready

timothybonci commented:
According to [this](https://github.com/npm/cli/issues/4017#issuecomment-1101481619) the design intent is to omit tagging when versioning workspaces.  We can still create commits.  I'll attempt to add a new condition before tagging that checks if we're in a workspace.  This will require good tests, too.

timothybonci committed:
Setup git user in CI tests

timothybonci commented:
Having gotten more familiar with this, I'm leaning toward detecting workspaces, and then doing no tag and no commit.  The convention we use in --git-tag-version false is "no tag, no commit".  Inventing a new halfway commit with no tag for this use case doesn't seem wise.

timothybonci committed:
Patch 2 (#1)

* fix: look parent dirs to detect git repo on version

* Mock find and test with subdirectories

* refactor: use let instead of var

* style: use single quotes

* Revert "refactor: use let instead of var"

This reverts commit 2e40ec9e52f02caca5d4c15e17a8d99fe7164b5c.

* use same directory in the setup and the tests

Fixing my silly mistake.  This fixes all but the last test.

* Fix final test

Last test needs subdirectory evaluation

Co-authored-by: fernandopasik <fernando@pasik.com.ar>

timothybonci committed:
Check for workspace safety before doing git operations

timothybonci committed:
reset package in workspace test and use path.sep for Windows compat

timothybonci committed:
fix merge conflicts

timothybonci committed:
readd node_modules redirect

timothybonci commented:
I think I've got the workspace case covered now.  https://github.com/fernandopasik/cli/pull/3
As I say in that PR, I'm not that skilled in Node, so I may have missed something silly/obvious to a seasoned dev.

timothybonci committed:
fix typo and clarify the workspaces condition

timothybonci commented:
Something I thought about, but didn't implement - should --f always result in a commit, even in the workspaces case?  It seems like the universal "let you do otherwise unwise things" but I worry about how overloaded it could get, turning into a footgun.  I'd love to get feedback on this if it's desirable.

fernandopasik committed:
Merge pull request #3 from timothybonci/patch-3

Patch 3

fernandopasik commented:
@timothybonci almost all the work here is yours, please feel free to open a new PR with all the changes so I don't delay you.
If you do link it here and I will close my PR

timothybonci commented:
Thanks for starting this @fernandopasik.  I've opened https://github.com/npm/cli/pull/5442 as a fresh start with my proposed changes.

fernandopasik commented:
No problem, thanks for your contributions too.
